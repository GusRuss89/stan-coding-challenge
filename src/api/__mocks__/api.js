const sample = [
  {
    title: 'Wolf Creek',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    programType: 'series',
    images: {
      'Poster Art': {
        url:
          'https://streamcoimg-a.akamaihd.net/000/128/61/12861-PosterArt-ec32a81986a45eac7e080112075ab466.jpg',
        width: 1000,
        height: 1500
      }
    },
    releaseYear: 2016
  },
  {
    title: 'Danny Collins',
    description:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    programType: 'movie',
    images: {
      'Poster Art': {
        url:
          'https://streamcoimg-a.akamaihd.net/000/999/667/999667-PosterArt-65e6f88ede8bdcdad6d4d4f1549936ee.jpg',
        width: 1000,
        height: 1500
      }
    },
    releaseYear: 2014
  }
]

export const getTitles = jest.fn(() => {
  return new Promise(resolve => {
    process.nextTick(() => resolve(sample))
  })
})
