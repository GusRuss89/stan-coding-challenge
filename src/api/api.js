export const getTitles = async () => {
  const res = await fetch('/sample.json')
  if (!res.ok) throw new Error('Response not ok')
  const json = await res.json()
  return json.entries
}
