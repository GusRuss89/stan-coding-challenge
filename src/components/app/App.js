import React, { useReducer, useEffect } from 'react'
import { getTitles } from '../../api/api'
import { filterTitles } from '../../utils/utils'
import { initialState, reducer } from '../../state'
import Header from '../layout/Header'
import PageTitle from '../layout/PageTitle'
import Footer from '../layout/Footer'
import Titles from '../titles/Titles'
import MenuItem from '../menu/MenuItem'
import styles from '../../theme/App.module.css'

const App = () => {
  const [state, dispatch] = useReducer(reducer, initialState)

  useEffect(() => {
    async function fetchTitles() {
      try {
        const allTitles = await getTitles()
        const filteredTitles = filterTitles(allTitles)
        dispatch({ type: 'update-titles', ...filteredTitles })
      } catch (e) {
        dispatch({ type: 'error', error: 'Oops, something went wrong...' })
      }
    }
    fetchTitles()
  }, [])

  const handleBack = () => dispatch({ type: 'navigate', display: 'menu' })

  const pageTitleMap = {
    menu: 'Popular Titles',
    series: 'Popular Series',
    movies: 'Popular Movies'
  }

  return (
    <div className={styles.root}>
      <div className={styles.header}>
        <Header />
      </div>
      <main className={styles.main}>
        <PageTitle
          title={pageTitleMap[state.display]}
          onBack={state.display === 'menu' ? null : handleBack}
        />

        <div className={styles.content}>
          <div className="page-width">
            <div className={styles.contentInner}>
              {state.status === 'loading' && (
                <span data-testid="loading" className={styles.loading}>
                  Loading...
                </span>
              )}

              {state.status === 'success' && (
                <div data-testid="successcontent">
                  {state.display === 'menu' ? (
                    <div data-testid="menu" className={styles.menu}>
                      <MenuItem
                        imgText="Series"
                        title="Popular Series"
                        onClick={() => dispatch({ type: 'navigate', display: 'series' })}
                      />
                      <MenuItem
                        imgText="Movies"
                        title="Popular Movies"
                        onClick={() => dispatch({ type: 'navigate', display: 'movies' })}
                      />
                    </div>
                  ) : (
                    <div data-testid="titles">
                      <Titles titles={state[state.display]} />
                    </div>
                  )}
                </div>
              )}

              {!['success', 'loading'].includes(state.status) && (
                <span data-testid="error" className={styles.error}>
                  {state.error || 'An unknown error occured...'}
                </span>
              )}
            </div>
          </div>
        </div>
      </main>
      <div className={styles.footer}>
        <Footer />
      </div>
    </div>
  )
}

export default App
