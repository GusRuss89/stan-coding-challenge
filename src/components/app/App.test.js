import React from 'react'
import renderer from 'react-test-renderer'
import { render, fireEvent, waitForElement, cleanup } from 'react-testing-library'
import App from './App'
import { getTitles } from '../../api/api'

jest.mock('../../api/api')
afterEach(cleanup)

function awaitMock() {
  const results = getTitles.mock.results
  const result = results[results.length - 1]
  if (result.isThrow) return Promise.resolve()
  return result.value
}

describe('App', () => {
  it('renders without crashing', done => {
    render(<App />)
    awaitMock().then(res => done())
  })

  it('passes snapshot', () => {
    const tree = renderer.create(<App />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('starts in loading state', done => {
    const { findByTestId } = render(<App />)
    expect(findByTestId('loading')).toBeTruthy()
    awaitMock().then(res => done())
  })

  it('fetches titles successfully', done => {
    const { getByTestId } = render(<App />)
    expect.assertions(2)
    expect(getTitles).toHaveBeenCalled()
    expect(async () => {
      await waitForElement(() => getByTestId('successcontent'))
    }).not.toThrow()
    awaitMock().then(res => done())
  })

  it('updates content on navigation', async done => {
    const { getByText, getByTestId } = render(<App />)
    await waitForElement(() => getByText('Popular Series'))
    fireEvent.click(getByText('Popular Series'))
    expect(async () => {
      await waitForElement(() => getByTestId('titles'))
    }).not.toThrow()
    awaitMock().then(res => done())
  })

  it('handles errors gracefully', done => {
    getTitles.mockImplementation(() => {
      throw new Error()
    })
    const { getByTestId } = render(<App />)
    expect.assertions(2)
    expect(getTitles).toHaveBeenCalled()
    expect(async () => {
      await waitForElement(() => getByTestId('error'))
    }).not.toThrow()
    awaitMock().then(res => done())
  })
})
