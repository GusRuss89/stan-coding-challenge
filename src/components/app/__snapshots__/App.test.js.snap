// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App passes snapshot 1`] = `
<div
  className="root"
>
  <div
    className="header"
  >
    <header
      className="root"
    >
      <div
        className="page-width"
      >
        <div
          className="inner"
        >
          <h1
            className="logo"
          >
            DEMO Streaming
          </h1>
          <nav
            className="nav"
          >
            <ul
              className="navList"
            >
              <li>
                <a
                  href="#login"
                >
                  Log in
                </a>
              </li>
              <li
                className="navButton"
              >
                <a
                  href="#trial"
                >
                  Start your free trial
                </a>
              </li>
            </ul>
          </nav>
        </div>
      </div>
    </header>
  </div>
  <main
    className="main"
  >
    <header
      className="root"
    >
      <div
        className="page-width"
      >
        <div
          className="inner"
        >
          <h1
            className="title"
          >
            Popular Titles
          </h1>
        </div>
      </div>
    </header>
    <div
      className="content"
    >
      <div
        className="page-width"
      >
        <div
          className="contentInner"
        >
          <span
            className="loading"
            data-testid="loading"
          >
            Loading...
          </span>
        </div>
      </div>
    </div>
  </main>
  <div
    className="footer"
  >
    <footer
      className="root"
    >
      <div
        className="page-width"
      >
        <div
          className="inner"
        >
          <nav
            className="nav"
          >
            <ul>
              <li>
                <a
                  href="#home"
                >
                  Home
                </a>
              </li>
              <li>
                <a
                  href="#terms"
                >
                  Terms and Conditions
                </a>
              </li>
              <li>
                <a
                  href="#privacy"
                >
                  Privacy Policy
                </a>
              </li>
              <li>
                <a
                  href="#collection"
                >
                  Collection Statement
                </a>
              </li>
              <li>
                <a
                  href="#help"
                >
                  Help
                </a>
              </li>
              <li>
                <a
                  href="#manage"
                >
                  Manage Account
                </a>
              </li>
            </ul>
          </nav>
          <p
            className="copyright"
          >
            Copyright Â© 
            2019
             DEMO Streaming. All rights reserved.
          </p>
          <div
            className="icons"
          >
            <ul
              className="social"
            >
              <li>
                <a
                  href="https://facebook.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    facebook-white.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://twitter.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    twitter-white.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://instagram.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    instagram-white.svg
                  </svg>
                </a>
              </li>
            </ul>
            <ul
              className="appStores"
            >
              <li>
                <a
                  href="https://apple.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    app-store.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  href="https://google.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    play-store.svg
                  </svg>
                </a>
              </li>
              <li>
                <a
                  className="windowsStoreIcon"
                  href="https://microsoft.com"
                  rel="noreferrer noopener"
                >
                  <svg>
                    windows-store.svg
                  </svg>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
    </footer>
  </div>
</div>
`;
